File: dotnet/jaeger/Example.Clients/test-OTEL-client/SimpleTest.cs
namespace test_OTEL_client;

public static class SimpleTest
{
    // Run simple test method
    public static async Task Run(IServiceProvider services, CancellationToken cancellationToken)
    {
        // Create a scope to get scoped services
        using IServiceScope serviceScope = services.CreateScope();

        // Get the WeatherClient, ILogger and ActivitySource services
        WeatherClient weatherClient = serviceScope.ServiceProvider.GetRequiredService<WeatherClient>();
        ILogger<WeatherClient> logger = serviceScope.ServiceProvider.GetRequiredService<ILogger<WeatherClient>>();
        ActivitySource activitySource = serviceScope.ServiceProvider.GetRequiredService<ActivitySource>();

        // Call the private method to get weather forecasts
        await GetWeatherForecastAsync(weatherClient, logger, activitySource, cancellationToken);
        // Create and update a weather forecast
        await CreateWeatherForecastAsync(weatherClient, logger, activitySource, cancellationToken);
        // Call the private method to get weather forecasts again
        await UpdateWeatherForecastAsync(weatherClient, logger, activitySource, cancellationToken);
    }

    // Run more complex test that calls method multiple times, concurrently, etc.
    public static async Task RunComplex(IServiceProvider services, CancellationToken cancellationToken)
    {
        // Create a scope to get scoped services
        using IServiceScope serviceScope = services.CreateScope();

        // Get the WeatherClient, ILogger and ActivitySource services
        WeatherClient weatherClient = serviceScope.ServiceProvider.GetRequiredService<WeatherClient>();
        ILogger<WeatherClient> logger = serviceScope.ServiceProvider.GetRequiredService<ILogger<WeatherClient>>();
        ActivitySource activitySource = serviceScope.ServiceProvider.GetRequiredService<ActivitySource>();

        // Call the private method to get weather forecasts multiple times
        List<Task> initialTasks = new();
        for (int i = 0; i < 3; i++) {
            initialTasks.Add(GetWeatherForecastAsync(weatherClient, logger, activitySource, cancellationToken));
        }

        // Call the private method to create and update weather forecasts concurrently
        List<Task> tasks = new();
        for (int i = 0; i < 5; i++) {
            tasks.Add(CreateWeatherForecastAsync(weatherClient, logger, activitySource, cancellationToken));
            tasks.Add(UpdateWeatherForecastAsync(weatherClient, logger, activitySource, cancellationToken));
        }
        await Task.WhenAll(tasks);
        await Task.WhenAll(initialTasks);

        // Call the private method to get weather forecasts again
        await GetWeatherForecastAsync(weatherClient, logger, activitySource, cancellationToken);
    }

    // private static method to get weather forecast
    private static async Task GetWeatherForecastAsync(WeatherClient weatherClient,
                                                        ILogger<WeatherClient> logger,
                                                        ActivitySource activitySource,
                                                        CancellationToken cancellationToken)
    {
        // Start the activity
        using Activity? activity = activitySource.StartActivity("GetWeatherForecast");

        try {
            // Call the WeatherClient to get weather forecasts
            WeatherForecastsDto forecasts = await weatherClient.GetWeatherForecastAsync(5, cancellationToken);

            // Set some tags to the activity
            activity?.SetTag("weather.forecast.count", forecasts.Count);

            // Set activity status
            activity?.SetStatus(ActivityStatusCode.Ok);

            // Log the result
            logger.LogInformation($"Retrieved {forecasts.Count} weather forecasts (out of {forecasts.AllCount} total):");
        } catch (Exception ex) {
            // Set activity status to error
            activity?.SetStatus(ActivityStatusCode.Error, ex.Message);

            // Log the error
            logger.LogError(ex, $"Error retrieving weather forecasts.");
        }
    }

    // private static method to set weather forecast
    private static async Task UpdateWeatherForecastAsync(WeatherClient weatherClient,
                                                        ILogger<WeatherClient> logger,
                                                        ActivitySource activitySource,
                                                        CancellationToken cancellationToken)
    {
        // Start the activity
        using Activity? activity = activitySource.StartActivity("UpdateWeatherForecast");

        // Get random weather forecast temperature
        int tempC = Random.Shared.Next(-20, -5);

        try {
            // Create a new weather forecast
            WeatherForecastDto newForecast = new(new DateOnly(2025, 12, 17), tempC, "Cold");

            // Call the WeatherClient to set the weather forecast
            WeatherForecastDto updatedForecast = await weatherClient.UpdateWeatherForecastAsync(newForecast, cancellationToken);

            // Set some tags to the activity
            activity?.SetTag("weather.forecast.date", updatedForecast.Date.ToString());
            activity?.SetTag("weather.forecast.temperatureC", updatedForecast.TemperatureC);
            activity?.SetTag("weather.forecast.summary", updatedForecast.Summary);

            // Set activity status
            activity?.SetStatus(ActivityStatusCode.Ok);

            // Log the result
            logger.LogInformation($"Set weather forecast for {updatedForecast.Date}: {updatedForecast.TemperatureC}C, {updatedForecast.Summary}");
        } catch (Exception ex) {
            // Set activity status to error
            activity?.SetStatus(ActivityStatusCode.Error, ex.Message);

            // Log the error
            logger.LogError(ex, $"Error setting weather forecast.");
        }
    }

    // private static method to create weather forecast
    private static async Task CreateWeatherForecastAsync(WeatherClient weatherClient,
                                                        ILogger<WeatherClient> logger,
                                                        ActivitySource activitySource,
                                                        CancellationToken cancellationToken)
    {
        // Start the activity
        using Activity? activity = activitySource.StartActivity("CreateWeatherForecast");

        // Get random weather forecast temperature
        int tempC = Random.Shared.Next(-20, -5);

        try {
            // Create a new weather forecast
            WeatherForecastDto newForecast = new(new DateOnly(2025, 12, 17), tempC, "Cold");

            // Call the WeatherClient to create the weather forecast
            WeatherForecastDto createdForecast = await weatherClient.CreateWeatherForecastAsync(newForecast, cancellationToken);

            // Set some tags to the activity
            activity?.SetTag("weather.forecast.date", createdForecast.Date.ToString());
            activity?.SetTag("weather.forecast.temperatureC", createdForecast.TemperatureC);
            activity?.SetTag("weather.forecast.summary", createdForecast.Summary);

            // Set activity status
            activity?.SetStatus(ActivityStatusCode.Ok);

            // Log the result
            logger.LogInformation($"Set weather forecast for {createdForecast.Date}: {createdForecast.TemperatureC}C, {createdForecast.Summary}");
        } catch (Exception ex) {
            // Set activity status to error
            activity?.SetStatus(ActivityStatusCode.Error, ex.Message);

            // Log the error
            logger.LogError(ex, $"Error setting weather forecast.");
        }
    }
}

File: dotnet/aspire/Example.Clients/test-OTEL-client-ts/src/simpleTest.ts
import { WeatherClient } from './weatherClient';
import { WeatherForecastDto, WeatherForecastsDto } from './types';
import { Span, trace, Tracer } from '@opentelemetry/api';
import { DefaultValues } from './defaultValues';

export class SimpleTest {

  private static tracer: Tracer = trace.getTracer(DefaultValues.ACTIVITY_SOURCE_NAME);

  // Run simple test method
  public static async run(): Promise<void> {

    // Get Weather Client
    const weatherClient = new WeatherClient();

    // Start a parent span
    SimpleTest.tracer.startActiveSpan('get-weather-forecast', async (parentSpan: Span) => {
      try {
        // Call the private method to get weather forecasts
        await this.getWeatherForecastAsync(weatherClient);
      } finally {
        // End parent span
        parentSpan.end();
      }
    });

    // Start a parent span
    SimpleTest.tracer.startActiveSpan('create-weather-forecast', async (parentSpan: Span) => {
      try {
        // Call the private method to create weather forecasts
        await this.createWeatherForecastAsync(weatherClient);
      } finally {
        // End parent span
        parentSpan.end();
      }
    });

    // Start a parent span
    SimpleTest.tracer.startActiveSpan('update-weather-forecast', async (parentSpan: Span) => {
      try {
        // Call the private method to update weather forecasts
        await this.updateWeatherForecastAsync(weatherClient);
      } finally {
        // End parent span
        parentSpan.end();
      }
    });
  }

  // Run more complex test that calls method multiple times, concurrently, etc.
  public static async runComplex(): Promise<void> {
    const weatherClient = new WeatherClient();

    // Call the private method to get weather forecasts multiple times
    const initialTasks: Promise<void>[] = [];
    for (let i = 0; i < 3; i++) {
      initialTasks.push(// Start a parent span
        SimpleTest.tracer.startActiveSpan('get-weather-forecast', async (parentSpan: Span) => {
          try {
            // Call the private method to get weather forecasts
            await this.getWeatherForecastAsync(weatherClient);
          } finally {
            // End parent span
            parentSpan.end();
          }
        }));
    }

    // Call the private method to create and update weather forecasts concurrently
    const tasks: Promise<void>[] = [];
    for (let i = 0; i < 5; i++) {
      tasks.push(// Start a parent span
        SimpleTest.tracer.startActiveSpan('create-weather-forecast', async (parentSpan: Span) => {
          try {
            // Call the private method to create weather forecasts
            await this.createWeatherForecastAsync(weatherClient);
          } finally {
            // End parent span
            parentSpan.end();
          }
        }));
      tasks.push(// Start a parent span
        SimpleTest.tracer.startActiveSpan('update-weather-forecast', async (parentSpan: Span) => {
          try {
            // Call the private method to update weather forecasts
            await this.updateWeatherForecastAsync(weatherClient);
          } finally {
            // End parent span
            parentSpan.end();
          }
        }));
    }
    await Promise.all(tasks);
    await Promise.all(initialTasks);
  }

  // private static method to get weather forecast
  private static async getWeatherForecastAsync(weatherClient: WeatherClient): Promise<void> {
    try {
      // Call the WeatherClient to get weather forecasts
      const forecasts: WeatherForecastsDto = await weatherClient.getWeatherForecastAsync(5);
      // Log the result
      console.log(`Retrieved ${forecasts.count} weather forecasts (out of ${forecasts.allCount} total):`);
    } catch (error) {
      // Log the error
      console.error('Error retrieving weather forecasts:', error);
    }
  }

  // private static method to set weather forecast
  private static async updateWeatherForecastAsync(weatherClient: WeatherClient): Promise<void> {
    // Get random weather forecast temperature
    const tempC = Math.floor(Math.random() * 15) - 20; // -20 to -5

    try {
      // Create a new weather forecast
      const newForecast: WeatherForecastDto = {
        date: '2025-12-17',
        temperatureC: tempC,
        summary: 'Cold'
      };

      // Call the WeatherClient to set the weather forecast
      const updatedForecast = await weatherClient.updateWeatherForecastAsync(newForecast);

      // Log the result
      console.log(`Updated weather forecast for ${updatedForecast.date}: ${updatedForecast.temperatureC}C, ${updatedForecast.summary}`);
    } catch (error) {
      // Log the error
      console.error('Error updating weather forecast:', error);
    }
  }

  // private static method to create weather forecast
  private static async createWeatherForecastAsync(weatherClient: WeatherClient): Promise<void> {
    // Get random weather forecast temperature
    const tempC = Math.floor(Math.random() * 15) - 20; // -20 to -5

    try {
      // Create a new weather forecast
      const newForecast: WeatherForecastDto = {
        date: '2025-12-17',
        temperatureC: tempC,
        summary: 'Cold'
      };

      // Call the WeatherClient to create the weather forecast
      const createdForecast = await weatherClient.createWeatherForecastAsync(newForecast);

      // Log the result
      console.log(`Created weather forecast for ${createdForecast.date}: ${createdForecast.temperatureC}C, ${createdForecast.summary}`);
    } catch (error) {
      // Log the error
      console.error('Error creating weather forecast:', error);
    }
  }
}